src/hotspot/share/opto/subnode.cpp

```diff
@@ -1814,100 +1893,10 @@ Node *BoolNode::Ideal(PhaseGVN *phase, bool can_reshape) {
-//      x       <u 0
-//   which is false for any x (Lemma 3)
-//
-// (2) Change ((x & (m - 1)) <u m) or (((m - 1) & x) <u m) to (m >u 0)
-// This is the off-by-one variant of the above.
-//
-// We now prove that this replacement is correct. This is the same as proving
-//   "m >u 0" if and only if "x & (m - 1) <u m", i.e. "m >u 0 <=> x & (m - 1) <u m"
-//
-// We use (Lemma 1) and (Lemma 3) from above.
-//
-// Case "x & (m - 1) <u m => m >u 0":
-//   We prove this by contradiction:
-//     Assume m <=u 0 which is equivalent to m == 0:
-//   and thus
-//     x & (m - 1) <u m = 0               // m == 0
-//     y           <u     0               // y = x & (m - 1)
-//   by Lemma 3, this is always false, i.e. a contradiction to our assumption.
-//
-// Case "m >u 0 => x & (m - 1) <u m":
-//   x & (m - 1) <=u (m - 1)              // (Lemma 1)
-//   x & (m - 1) <=u (m - 1) <u m         // Using assumption m >u 0, no underflow of "m - 1"
-//
-//
-// Note that the signed version of "m > 0":
-//   m > 0 <=> x & (m - 1) <u m
-// does not hold:
-//   Assume m == -1 and x == -1:
-//     x  & (m - 1) <u m
-//     -1 & -2      <u -1
-//     -2           <u -1
-//     UINT_MAX - 1 <u UINT_MAX           // Signed to unsigned numbers
-// which is true while
-//   m > 0
-// is false which is a contradiction.
-//
-// (1a) and (1b) is covered by this method since we can directly return a true value as type while (2) is covered
-// in BoolNode::Ideal since we create a new non-constant node (see [CMPU_MASK]).
-const Type* BoolNode::Value_cmpu_and_mask(PhaseValues* phase) const {
-  Node* cmp = in(1);
-  if (cmp != nullptr && cmp->Opcode() == Op_CmpU) {
-    Node* cmp1 = cmp->in(1);
-    Node* cmp2 = cmp->in(2);
-
-    if (cmp1->Opcode() == Op_AndI) {
-      Node* m = nullptr;
-      if (_test._test == BoolTest::le) {
-        // (1a) "((x & m) <=u m)", cmp2 = m
-        m = cmp2;
-      } else if (_test._test == BoolTest::lt && cmp2->Opcode() == Op_AddI && cmp2->in(2)->find_int_con(0) == 1) {
-        // (1b) "(x & m) <u m + 1" and "(m & x) <u m + 1", cmp2 = m + 1
-        Node* rhs_m = cmp2->in(1);
-        const TypeInt* rhs_m_type = phase->type(rhs_m)->isa_int();
-        if (rhs_m_type != nullptr && (rhs_m_type->_lo > -1 || rhs_m_type->_hi < -1)) {
-          // Exclude any case where m == -1 is possible.
-          m = rhs_m;
-        }
-      }
-
-      if (cmp1->in(2) == m || cmp1->in(1) == m) {
-        return TypeInt::ONE;
-      }
-    }
-  }
-
-  return nullptr;
-}
-
 // Simplify a Bool (convert condition codes to boolean (1 or 0)) node,
 // based on local information.   If the input is constant, do it.
 const Type* BoolNode::Value(PhaseGVN* phase) const {
-  const Type* input_type = phase->type(in(1));
-  if (input_type == Type::TOP) {
-    return Type::TOP;
-  }
-  const Type* t = Value_cmpu_and_mask(phase);
-  if (t != nullptr) {
-    return t;
-  }
-
-  return _test.cc2logical(input_type);
+  return _test.cc2logical( phase->type( in(1) ) );
```

Comment:
```suggestion
  return _test.cc2logical(phase->type(in(1)));
```

Reply:
Suggestion accepted in 27ed1a311ec34e24afae6cc43d2c71e2620eb0ef.